syntax = "proto3";

/* Define package */
package pymerang;

/* Import status codes */
import "pymerang/status_codes.proto";


/* RPC methods implementing a procedure to connect a device to a controller */

service Pymerang {
  // Register a device
  rpc RegisterDevice(RegisterDeviceRequest) returns (RegisterDeviceReply) {}
  // Update management information
  rpc UpdateMgmtInfo(RegisterDeviceRequest) returns (RegisterDeviceReply) {}
  // Request to generate a certificate
  rpc SignCertificate(SignCertificateRequest) returns (SignCertificateReply) {}
  // Unregister device registration
  rpc UnregisterDevice(RegisterDeviceRequest) returns (RegisterDeviceReply) {}
}

/* Message representing a feature supported by a device */
message Feature {
  string name = 1;
  uint32 port = 2;
}

/* Authentication data */
message AuthData {
  string token = 1;
}

/* Message representing a device */
message Device {
    string id = 1;
    repeated Feature features = 2;
}

/* Management Info */
message MgmtInfo {
    string tunnel_mode = 1;
    string device_vtep_ip = 2;
    string controller_vtep_ip = 3;
    uint32 vtep_mask = 4;
    string device_external_ip = 5;
    uint32 device_external_port = 6;
    string device_vtep_mac = 8;
    string controller_vtep_mac = 9;
    uint32 vxlan_port = 10;
    string nat_type = 11;
}

/* Interface */
message Interface {
    string name = 1;
    string mac_addr = 2;
    repeated string ipv4_addrs = 3;
    repeated string ipv6_addrs = 4;
    repeated string ext_ipv4_addrs = 5;
    repeated string ext_ipv6_addrs = 6;
}

/* RegisterDevice request message */
message RegisterDeviceRequest {
  Device device = 1;
  AuthData auth_data = 2;
  MgmtInfo mgmt_info = 3;
  repeated Interface interfaces = 4;
  string tenantid = 5;
}

/* Reports the outcome of a gRPC request command */
message RegisterDeviceReply {
  StatusCode status = 1;
  MgmtInfo mgmt_info = 2;
  string tenantid = 3;
}

/* Request for a certificate */
message SignCertificateRequest {
  Device device = 1;
  bytes csr = 2;
  AuthData auth_data = 3;
}

/* Reply containing a certificate */
message SignCertificateReply {
  StatusCode status = 1;
  bytes cert = 2;
}